File: wind-composer/drum_engine.py
"""Step-sequenced drum generation."""

from __future__ import annotations

import random
from dataclasses import dataclass
from typing import List

from rhythm_engine import RhythmEvent
from style_engine import StyleProfile
from utils import clamp


@dataclass
class DrumContext:
    beat: int
    bar: int
    energy: float
    precipitation: float
    snowfall: float
    style: StyleProfile
    section_energy: float


class DrumEngine:
  def __init__(self) -> None:
    self._hat_toggle = False

  def on_beat(self, ctx: DrumContext) -> List[RhythmEvent]:
    events: List[RhythmEvent] = []
    step = ctx.beat % 16
    pattern = ctx.style.kick_pattern
    density = ctx.style.drum_density * (0.5 + ctx.energy * 0.5) * ctx.section_energy

    if pattern[step] and random.random() < 0.95:
      events.append(RhythmEvent(layer="kick", strength=clamp(0.5 + ctx.energy * 0.35), is_pulse=True))

    if density > 0.25 and step % 2 == 1:
      hat_str = clamp(density * 0.35 + random.uniform(0, 0.15))
      events.append(RhythmEvent(layer="hat", strength=hat_str, is_pulse=False))

    if density > 0.45 and step in (4, 12) and random.random() < density:
      events.append(RhythmEvent(layer="snare", strength=clamp(0.4 + ctx.energy * 0.3), is_pulse=True))

    if ctx.precipitation > 0.5 and step % 4 == 2:
      events.append(RhythmEvent(layer="percussion", strength=clamp(ctx.precipitation * 0.25), is_pulse=False))

    if ctx.snowfall > 0.2 and step % 8 == 0:
      events.append(RhythmEvent(layer="bell", strength=0.2, is_pulse=False))

    return events

File: wind-composer-web/src/music/drumEngine.ts
import type { RhythmEventDto } from "../types";
import { clamp, randBetween } from "../utils";
import type { StyleProfile } from "./styleEngine";

export interface DrumContext {
  beat: number;
  bar: number;
  energy: number;
  precipitation: number;
  snowfall: number;
  style: StyleProfile;
  sectionEnergy: number;
}

export class DrumEngine {
  onBeat(ctx: DrumContext): RhythmEventDto[] {
    const events: RhythmEventDto[] = [];
    const step = ctx.beat % 16;
    const pattern = ctx.style.kickPattern;
    const density = ctx.style.drumDensity * (0.5 + ctx.energy * 0.5) * ctx.sectionEnergy;

    if (pattern[step] && Math.random() < 0.96) {
      events.push({ layer: "kick", strength: clamp(0.55 + ctx.energy * 0.35), is_pulse: true });
    }

    const hatPattern = ctx.style.hatPattern;
    if (density > 0.12 && hatPattern[step]) {
      const vel = clamp(density * 0.4 + randBetween(0, 0.12));
      const layer = ctx.style.offbeatOpenHat && step % 4 === 2 ? "open_hat" : "hat";
      events.push({ layer, strength: vel, is_pulse: false });
    }

    if (density > 0.2 && step % 2 === 1 && Math.random() < density * 0.85) {
      events.push({ layer: "hat", strength: clamp(density * 0.28), is_pulse: false });
    }

    if (density > 0.35 && (step === 4 || step === 12) && Math.random() < density) {
      const layer = ctx.style.useClap && step === 12 ? "clap" : "snare";
      events.push({ layer, strength: clamp(0.45 + ctx.energy * 0.35), is_pulse: true });
    }

    if (ctx.style.name === "Trance" && density > 0.5 && step % 2 === 0) {
      events.push({ layer: "hat", strength: clamp(density * 0.32), is_pulse: false });
    }

    if (ctx.style.name === "Deep House" && density > 0.4 && step % 4 === 2) {
      events.push({ layer: "percussion", strength: clamp(density * 0.35), is_pulse: false });
    }

    if (ctx.precipitation > 0.5 && step % 4 === 2) {
      events.push({ layer: "noise", strength: clamp(ctx.precipitation * 0.22), is_pulse: false });
    }

    if (ctx.snowfall > 0.2 && step % 8 === 0) {
      events.push({ layer: "ride", strength: 0.18, is_pulse: false });
    }

    if (Math.random() < density * 0.04 * ctx.energy) {
      events.push({ layer: "hat", strength: 0.12, is_pulse: false });
    }

    return events;
  }
}
